python/tank/bootstrap/manager.py

```diff
@@ -236,17 +298,62 @@ def bootstrap_engine(self, engine_name, entity=None):
         the engine may not be the same as the API version that was
         executed during the bootstrap.
 
-        :param entity: Shotgun entity to launch engine for
-        :type entity: Dictionary with keys type and id
+        :param engine_name: Name of engine to launch (e.g. ``tk-nuke``).
+        :param entity: Shotgun entity to launch engine for.
+        :param progress_callback: Callback function that reports back on the toolkit and engine bootstrap progress.
+        :param completed_callback: Callback function that handles cleanup after successful completion of the bootstrap.
+        :param failed_callback: Callback function that handles cleanup after failed completion of the bootstrap.
+        :returns: :class:`sgtk.platform.Engine` instance when ``async_bootstrap` is false; otherwise None.
+        """
+
+        log.info("Bootstrapping engine %s for entity %s." % (engine_name, entity))
+
+        bootstrap_asynchronously = self.async_bootstrap
+
+        if bootstrap_asynchronously:
+            try:
+                import threading
+            except ImportError:
+                log.warning("Cannot bootstrap asynchronously in a background thread.")
+                # Since Qt is not available, fall back on synchronous boostraping.
+                bootstrap_asynchronously = False
+
+        if bootstrap_asynchronously:
+
+            # Bootstrap an Sgtk instance asynchronously in a background thread,
+            # followed by launching the engine synchronously in the main application thread.
+            self._bootstrapper = threading.BootstrapSupervisor(self, engine_name, entity)
+            self._bootstrapper.set_callbacks(progress_callback, completed_callback, failed_callback)
+            self._bootstrapper.bootstrap()
+
+            # At this point, the engine has not yet been started.
+            engine = None
+
+        else:
+
+            # Execute the whole engine bootstrap logic synchronously in the main application thread.
+            tk = self.bootstrap_sgtk(engine_name, entity)
+            engine = self.start_engine(tk, engine_name, entity)
+
+        return engine
+
+    def start_engine(self, tk, engine_name, entity):
```

Comment:
I'm going to channel @manneohrstrom here, but we should be mindful of the methods we expose on the manager. This exposes `start_engine` and `bootstrap_sgtk` as public methods. I'm wondering if there's a true need to make these public other than to make it reasonable for the bootstrap thread class to call. 

That said, I have a situation in some code I'm working on where I would like to separate the bootstrapping of toolkit and the engine startup. It would be good to get Manne's input on this before merging.